sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -159,8 +159,9 @@ case class DataSourceAnalysis(conf: CatalystConf) extends Rule[LogicalPlan] {
 
 
     case i @ logical.InsertIntoTable(
-           l @ LogicalRelation(t: HadoopFsRelation, _, _), part, query, overwrite, false)
-        if query.resolved && t.schema.asNullable == query.schema.asNullable =>
```

Comment:
Did an investigation. For the struct type, we store the nested schema in the data type. Thus, after we cast the data type of struct, we also copy the schema at the same time. 

Below is the explained plan of the example I just added.
```
== Analyzed Logical Plan ==
InsertIntoHadoopFsRelationCommand file:/Users/xiaoli/IdeaProjects/sparkDelivery/spark-warehouse/tab2, ParquetFormat, <function0>, Map(path -> file:/Users/xiaoli/IdeaProjects/sparkDelivery/spark-warehouse/tab2), Append
   +- Project [cast(s#36 as struct<c:string,d:string>) AS p#40]
      +- Project [s#36]
         +- SubqueryAlias tab1
            +- Relation[s#36] parquet
```